README.md
# LoopBack

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/strongloop/loopback?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Module LTS Adopted'](https://img.shields.io/badge/Module%20LTS-Adopted-brightgreen.svg?style=flat)](http://github.com/CloudNativeJS/ModuleLTS)
[![IBM Support](https://img.shields.io/badge/IBM%20Support-Frameworks-brightgreen.svg?style=flat)](http://ibm.biz/node-support)

**⚠️ LoopBack 3 is in Maintenance LTS mode, only critical bugs and critical
security fixes will be provided. (See
[Module Long Term Support Policy](#module-long-term-support-policy) below.)**

We urge all LoopBack 3 users to migrate their applications to LoopBack 4 as
soon as possible. Refer to our
[Migration Guide](https://loopback.io/doc/en/lb4/migration-overview.html)
for more information on how to upgrade.

## Overview

LoopBack is a highly-extensible, open-source Node.js framework that enables you to:

  * Create dynamic end-to-end REST APIs with little or no coding.
  * Access data from Oracle, MySQL, PostgreSQL, MS SQL Server, MongoDB, SOAP and other REST APIs.
  * Incorporate model relationships and access controls for complex APIs.
  * Use built-in push, geolocation, and file services for mobile apps.
  * Easily create client apps using Android, iOS, and JavaScript SDKs.
  * Run your application on-premises or in the cloud.

LoopBack consists of:

  * A library of Node.js modules.
  * [Yeoman](http://yeoman.io/) generators for scaffolding applications.
  * Client SDKs for iOS, Android, and web clients.

LoopBack tools include:
  * Command-line tool `loopback-cli` to create applications, models, data sources, and so on.

For more details, see [https://loopback.io/](https://loopback.io/).


## Module Long Term Support Policy

LoopBack 3.x is now in Active LTS.

This module adopts the [Module Long Term Support (LTS)](http://github.com/CloudNativeJS/ModuleLTS) policy, with the following End Of Life (EOL) dates:

| Version    | Status          | Published | EOL                  |
| ---------- | --------------- | --------- | -------------------- |
| LoopBack 4 | Current         | Oct 2018  | Apr 2023 _(minimum)_ |
| LoopBack 3 | Maintenance LTS | Dec 2016  | Dec 2020             |
| LoopBack 2 | End-of-Life     | Jul 2014  | Apr 2019             |

Learn more about our LTS plan in [docs](https://loopback.io/doc/en/contrib/Long-term-support.html).

## LoopBack modules

The LoopBack framework is a set of Node.js modules that you can use independently or together.

![LoopBack modules](https://github.com/strongloop/loopback/raw/master/docs/assets/lb-modules.png "LoopBack modules")

### Core
* [loopback](https://github.com/strongloop/loopback)
* [loopback-datasource-juggler](https://github.com/strongloop/loopback-datasource-juggler)
* [strong-remoting](https://github.com/strongloop/strong-remoting)

### Connectors
* [loopback-connector-mongodb](https://github.com/strongloop/loopback-connector-mongodb)
* [loopback-connector-mysql](https://github.com/strongloop/loopback-connector-mysql)
* [loopback-connector-postgresql](https://github.com/strongloop/loopback-connector-postgresql)
* [loopback-connector-rest](https://github.com/strongloop/loopback-connector-rest)

### Enterprise Connectors
* [loopback-connector-oracle](https://github.com/strongloop/loopback-connector-oracle)
* [loopback-connector-mssql](https://github.com/strongloop/loopback-connector-mssql)
* [loopback-connector-soap](https://github.com/strongloop/loopback-connector-soap)
* [loopback-connector-atg](https://github.com/strongloop/loopback-connector-atg)

### Community Connectors

The LoopBack community has created and supports a number of additional connectors.  See [Community connectors](https://loopback.io/doc/en/lb2/Community-connectors.html) for details.

### Components
* [loopback-component-push](https://github.com/strongloop/loopback-component-push)
* [loopback-component-storage](https://github.com/strongloop/loopback-component-storage)
* [loopback-component-passport](https://github.com/strongloop/loopback-component-passport)

### Client SDKs
* [loopback-sdk-ios](https://github.com/strongloop/loopback-sdk-ios)
* [loopback-sdk-android](https://github.com/strongloop/loopback-sdk-android)
* [loopback-sdk-angular](https://github.com/strongloop/loopback-sdk-angular)
  * [loopback-sdk-angular-cli](https://github.com/strongloop/loopback-sdk-angular-cli)
  * [grunt-loopback-sdk-angular](https://github.com/strongloop/grunt-loopback-sdk-angular)

### Tools
* [loopback-explorer](https://github.com/strongloop/loopback-explorer)
* [loopback-workspace](https://github.com/strongloop/loopback-workspace)
* [generator-loopback](https://github.com/strongloop/generator-loopback)

### Examples

StrongLoop provides a number of example applications that illustrate various key LoopBack features. In some cases, they have accompanying step-by-step instructions (tutorials).

See [examples at loopback.io](https://loopback.io/examples/) for details.

## Resources

  * [Documentation](https://loopback.io/doc/).
  * [API documentation](https://apidocs.strongloop.com/loopback).
  * [LoopBack Announcements](https://groups.google.com/forum/#!forum/loopbackjs-announcements)
  * [LoopBack Google Group](https://groups.google.com/forum/#!forum/loopbackjs).
  * [GitHub issues](https://github.com/strongloop/loopback/issues).
  * [Gitter chat](https://gitter.im/strongloop/loopback).

## Contributing

Contributions to the LoopBack project are welcome! See [Contributing to LoopBack](https://loopback.io/doc/en/contrib/index.html) for more information.

## Reporting issues

One of the easiest ways to contribute to LoopBack is to report an issue. See [Reporting issues](https://loopback.io/doc/en/contrib/Reporting-issues.html) for more information.

[![Analytics](https://sl-beacon.appspot.com/UA-37775386-1/github/loopback/readme?pixel)](https://github.com/strongloop/loopback)


CHANGES.md
2020-11-25, Version 3.28.0
==========================

 * upgrade nodemailer to greater than 6.4.16 (jannyHou)

 * chore: sync LoopBack 4 with Node.js v14 EOL (Rifa Achrinza)

 * chore: add Node.js 14 to travis (Diana Lau)

 * Remove "major" and "p1" from stalebot (Miroslav Bajtoš)


2020-03-06, Version 3.27.0
==========================

 * Update LTS status in README (Miroslav Bajtoš)

 * chore: update copyright year (Diana Lau)

 * feat: change hasone relation error message (Sujesh T)

 * chore: disable security issue reporting (Nora)

 * chore: fix eslint violations (Nora)

 * fixup! manual fixes (Miroslav Bajtoš)

 * fixup! eslint --fix . (Miroslav Bajtoš)

 * chore: update eslint & eslint-config to latest (Miroslav Bajtoš)

 * chore: update dev-dependencies (Miroslav Bajtoš)

 * chore: update chai to v4, dirty-chai to v2 (Miroslav Bajtoš)

 * Updated "ismail" package to v3.2 (Stanislav Sarbinski)

 * Introduce issue templates for bugs, features, etc. (Miroslav Bajtoš)

 * Improve PULL_REQUEST_TEMPLATE (Miroslav Bajtoš)

 * test: use Chromium (not Chrome) when available (Miroslav Bajtoš)

 * test: disable Chrome sandboxing when inside Docker (Miroslav Bajtoš)

 * test: switch from PhantomJS to HeadlessChrome (Miroslav Bajtoš)


2019-05-31, Version 3.26.0
==========================

 * fix: disallow queries in username and email fields (Hage Yaapa)

 * Ignore failing downstream dependencies (Miroslav Bajtoš)

 * Upgrade nyc to version 14 (Miroslav Bajtoš)

 * Update Karma dependencies to latest versions (Miroslav Bajtoš)

 * Drop Node.js 6.x from the supported versions (Miroslav Bajtoš)

 * Fix Model.exists() to work with remote connector (Maxim Sharai)

 * chore: update copyrights years (Agnes Lin)

 * Update LTS status (Diana Lau)

 * Enable Node.js 12.x on Travis CI (Miroslav Bajtoš)

 * chore: update copyright year (Diana Lau)

 * chore: update LB3 EOL date (Diana Lau)


2019-03-15, Version 3.25.1
==========================

 * Back-ticks added to highlight example JSON (Quentin Presley)

 * Add same change to description for findOne (Quentin Presley)

 * Update the description for persisted-models (Quentin Presley)

 * handle $2b$ in hashed password check (Sylvain Dumont)


2019-02-05, Version 3.25.0
==========================

 * Support middleware injected by AppDynamics. (Mike Li)


2019-01-11, Version 3.24.2
==========================

 * Fix crash when modifying an unknown user (Matheus Horstmann)


2019-01-08, Version 3.24.1
==========================

 * Update underscore.string to 3.3.5 (Francois)

 * Fix: treat empty access token string as undefined (andrey-abramow)


2018-11-15, Version 3.24.0
==========================

 * Set juggler options for remote calls (Raymond Feng)

 * Speed up ACL tests by reducing saltWorkFactor (Miroslav Bajtoš)


2018-10-25, Version 3.23.2
==========================

 * Fix ACL check to support model wildcard (Moshe Malka)


2018-10-18, Version 3.23.1
==========================

 * README: highlight Active LTS at the top (Miroslav Bajtoš)


2018-10-09, Version 3.23.0
==========================

 * Clear handler cache when a method is added/removed (Mohammed Essehemy)

 * Add `options.preserveAccessTokens` (lchaglla)

 * Update LB3 to be active LTS (Diana Lau)

 * Fix ACL tests to wait until all assertions finish (Moshe Malka)

 * chore: update to latest linting rules (virkt25)


2018-09-12, Version 3.22.3
==========================

 * chore: use grunt to install optional  phantomjs (virkt25)

 * [WebFM] fr translation (candytangnb)


2018-08-29, Version 3.22.2
==========================

 * [WebFM] tr translation (candytangnb)

 * [WebFM] de translation (candytangnb)

 * [WebFM] cs/es/fr/it/nl/pl/pt_BR/ru translation (candytangnb)


2018-08-22, Version 3.22.1
==========================

 * [WebFM] ja/ko/zh_CN/zh_TW translation (candytangnb)

 * remove unnecessary format call (Diana Lau)

 * Make desc when export-api-def translatable (Diana Lau)


2018-08-08, Version 3.22.0
==========================

 * fix: accessToken create default acl (virkt25)

 * add: ppc64 and s390x to not run UI tests (Thomas Leah)

 * chore: update deps + fix linting + .npmrc (virkt25)

 * Update Loopback 2.x EOL dates (Chris Bailey)

 * Fix formatting (Chris Bailey)

 * Update support badge and move LTS section (Chris Bailey)

 * Add badges and information for LTS and support (Chris Bailey)


2018-07-09, Version 3.21.0
==========================

 * Make verifyUserRelations() more robust (mcitdev)

 * Fix crash in verifyUserRelations (ryanxwelch)

 * Fix crash in User model's "before delete" hook (mcitdev)

 * [WebFM] cs/pl/ru translation (candytangnb)

 * Update strong-error-handler (shimks)


2018-06-12, Version 3.20.0
==========================

 * Update strong-globalize to 4.x (Miroslav Bajtoš)

 * Update nodemailer to v4.x (Dimitris)

 * Drop support for Node.js 4.x (Miroslav Bajtoš)


2018-06-04, Version 3.19.3
==========================

 * Provide link to CODEOWNERS (Aditya Agarwal)

 * fix bug in User.verify when confirm is disabled (wolrajhti)

 * Enable Node.js 10.x on Travis CI (Miroslav Bajtoš)


2018-05-29, Version 3.19.2
==========================

 * Add check for undefined user email in setter (Kevin Scroggins)


2018-05-21, Version 3.19.1
==========================

 * Fix isOwner() bug in multiple-principal setup (Miroslav Bajtoš)


2018-04-17, Version 3.19.0
==========================

 * feat: remove all references to a Model (Miroslav Bajtoš)


2018-03-22, Version 3.18.3
==========================

 * Remove forgotten debugger statement (Miroslav Bajtoš)

 * Fix role check in apps with multiple user models (Miroslav Bajtoš)

 * Fix formatting issues reported by recent eslint (Miroslav Bajtoš)

 * CODEOWNERS: add nitro404 (Miroslav Bajtoš)

 * test: add missing "return" in a promise-style test (Miroslav Bajtoš)


2018-02-08, Version 3.18.2
==========================

 * model: fix infinite loop on nestRemoting (Kevin Delisle)


changelog.rb
#!/usr/bin/env ruby

require 'active_support'
require 'active_support/core_ext'
require 'octokit'
require 'faraday-http-cache'
require 'typhoeus/adapters/faraday'
require 'base64'
require 'yaml'

repo_types = (ARGV[0] || 'public').split(',')
span = ARGV[1] || 'week'

today = Date.current
last_week = today.last_week

sprints = [
  [Date.parse('February 9, 2016'), Date.parse('February 22, 2016'), 3]
]
while sprints.last[1] < today do
  prev = sprints.last
  sprints << [prev[0] + 14.days, prev[1] + 14.days, prev[2] + 1]
end
weeks = sprints.last(4).reverse

gh_token = ENV['GITHUB_TOKEN']
if gh_token.nil? or gh_token.empty?
  gh_token = `git config github.token`.strip
end

if gh_token.empty?
  STDERR.puts 'Please visit https://github.com/settings/tokens/new to generate a token'
  STDERR.puts "Once you have your token, run 'git config github.token <token>'"
  exit 1
end

# Response middleware that overrides the max-age.
# This allows the caching middleware to store things for up to a week
class MoarCacheable < Faraday::Middleware
  def call(request_env)
    @app.call(request_env).on_complete do |response_env|
      if response_env[:response_headers]["cache-control"] == "private, max-age=60, s-maxage=60"
        response_env[:response_headers]["cache-control"] = "private, max-age=604800, s-maxage=604800"
      end
    end
  end
end

class SimpleLogger
  def initialize(stream)
    @stream = stream
  end
  def debug(*a)
  #  @stream.puts *a
  end
end

store = ActiveSupport::Cache.lookup_store(:file_store, "/tmp/gh_cache")

stack = Faraday::RackBuilder.new do |builder|
  builder.use Faraday::HttpCache, shared_cache: false, store: store, serializer: Marshal, logger: SimpleLogger.new(STDERR)
  builder.use Octokit::Response::RaiseError
  builder.adapter Faraday.default_adapter
end
#npm = Faraday.new(url: 'https://registry.npmjs.org', builder: stack)

gh = Octokit::Client.new(access_token: gh_token, auto_paginate: true, middleware: stack)
if not gh.user_authenticated?
  STDERR.puts "Unauthenticated connection to GitHub, private repos will be invisible"
end

gh_starting = gh.rate_limit!.remaining
STDERR.puts "GitHub API: #{gh.rate_limit}"

$commitsToSkip = Hash.new {|h,k| h[k] = []}

def maybe_pr(gh, path, number)
  gh.pull_request(path, number) rescue false
end

def tryto(tries=3, dflt=nil)
  yield
rescue
  retry unless (tries -= 1).zero?
  dflt
end

def pp_commit(gh, r, commit)
  msg = commit.commit.message.lines.first.strip
  if $commitsToSkip[r.full_name].include? commit.sha
    nil
  elsif msg =~ /\AMerge pull request #(\d+) from/ and pr = maybe_pr(gh, r.full_name, $~[1])
    tryto(2,[]){gh.pull_request_commits(r.full_name, pr.number)}.each do |c|
      $commitsToSkip[r.full_name] << c.sha
    end
    "[PR##{pr.number}](#{pr.html_url}) #{pr.title} ([#{pr.user.login}](#{pr.user.html_url}))"
  elsif msg =~ /\Av?\d+\.\d+\.\d+(:?-[a-z0-9]+)?\Z/
    c = commit
    gc = commit.commit # lower level git info
    author = gc.author.name
    if c.author
      author = "[#{author}](#{c.author.html_url})"
    end
    "**Released #{msg}** (#{author})"
  else
    c = commit
    gc = commit.commit # lower level git info
    author = gc.author.name
    if c.author
      author = "[#{author}](#{c.author.html_url})"
    end
    "[@#{c.sha[0..6]}](#{c.html_url}) #{msg} (#{author})"
  end
end

def badge(alt, img, url)
  "[![#{alt}](#{img})](#{url})"
end

def issues_badge(repo)
  badge('GitHub Issues',
        "https://img.shields.io/github/issues/#{repo.full_name}.svg",
        "#{repo.html_url}/issues")
end

def downloads_badge(repo)
  badge('Downloads',
        "https://img.shields.io/npm/dm/#{repo.npm_name}.svg",
        "https://npmjs.com/package/#{repo.npm_name}")
end

def pp_week(week, repos)
  repos.group_by { |r|
    if r.name =~ /connector/
      [1, 'Connectors']
    elsif r.name =~ /component/
      [3, 'Components']
    elsif r.name =~ /-sdk-/
      [2, 'SDKs']
    else
      [0, 'Core']
    end
  }.sort.each do |group, repos|
    puts "\n## #{group[1]}\n"
    repos.sort_by(&:name).each do |repo|
      commits = repo.commits.reject { |c|
                  c.commit.message =~ /\AMerge (branch|tag)/
                }.reject { |c|
                  c.commit.committer.date <= week[0] or
                  c.commit.committer.date >= week[1]
                }.map {|c|
                  pp_commit(repo.gh, repo, c)
                }.compact
      next if commits.empty?
      heading = repo.npm_name || repo.name
      #heading << " #{issues_badge(repo)}"
      #if repo.npm_json.nil?
      #  heading << ' *package not available on npm*'
      #else
      #  heading << " #{downloads_badge(repo)}"
      #end
      puts "\n### #{heading}\n"
      # puts "#### v.Next"
      commits.each do |c|
        if c =~ /\A#+/
          puts
          puts c
        else
          puts " * #{c}"
        end
      end
      puts
    end
    puts
  end
end

REPOS = [
  'loopback',
  'loopback-boot',
  'loopback-datasource-juggler',
  'loopback-sdk-angular',
  'loopback-sdk-ios',
  'loopback-sdk-angular-cli',
  'loopback-sdk-xamarin',
  'generator-loopback',
  'loopback-connector-*',
  'loopback-swagger',
  'loopback-component-*',
  'loopback-phase',
  'loopback-filters',
  'strong-remoting',
  'strong-task-emitter',
]

def get_repos(gh, orgs, repo_types, weeks)
  orgs.flat_map { |o|
    repo_types.flat_map { |t|
      gh.org_repos(o, type: t)
    }
  }.select { |r|
    REPOS.any? {|pat| File.fnmatch?(pat, r.name) }
  }.map { |r|
    r.tap { |r|
      r.gh = gh
      if r.name == 'strongops'
        r.pkg_json_c = gh.contents(r.full_name, path: 'agent/package.json') rescue nil
      else
        r.pkg_json_c = gh.contents(r.full_name, path: 'package.json') rescue nil
      end
      if r.pkg_json_c and r.pkg_json_c.type == 'file' and r.pkg_json_c.encoding == 'base64'
        r.pkg_json = Base64.decode64(r.pkg_json_c.content)
        r.pkg = YAML.load(r.pkg_json)
        r.npm_name = r.pkg['name']
        r.npm_json = npm.get("/#{r.npm_name}") rescue false
        r.npm_info = YAML.load(r.npm_json.body) rescue {}
        r.npm_releases = {}
        (r.npm_info['time'] || []).each do |ev, at|
          at = Time.parse(at)
          if ev != 'modified' and ev != 'created' and at >= weeks.last[0] and at <= weeks.first[1]
            r.npm_releases[ev] = at
          end
        end
      else
        r.pkg_json = nil
        r.npm_releases = {}
        r.npm_info = {}
      end
      if r.pkg_name == 'strong-agent'
        r.commits = gh.commits_between(r.full_name, weeks.last[0], weeks.first[1], path: 'agent')
      else
        r.commits = gh.commits_between(r.full_name, weeks.last[0], weeks.first[1]) rescue []
      end
      r.shas = r.commits.map(&:sha)
    }
  }
end

repos = get_repos(gh, ['strongloop'], repo_types, weeks)

STDERR.puts "[GH] gathering repos used: #{gh_starting - gh.rate_limit!.remaining} requests (#{gh.rate_limit})"

weeks = []
dates = [Date.current - 8.weeks, Date.current]

# Frontmatter for Jekyll
puts '---'
puts 'layout: page'
puts "since: (from #{dates[0]} to #{dates[1]})"
puts '---'


weeks.each do |week|
  heading = "Sprint #{week[2]} (#{week[0]} to #{week[1]})"
  if week[1].future?
    heading << ' so far'
  end
  puts "## #{heading}\n"
  pp_week(week, repos)
  STDERR.puts "[GH] post pp_week(#{week[0]}) used: #{gh_starting - gh.rate_limit!.remaining} requests (#{gh.rate_limit})"
  puts '----'
  puts
end

pp_week(dates, repos)

STDERR.puts "[GH] Used #{gh_starting - gh.rate_limit!.remaining} requests (#{gh.rate_limit})"
